src/wccp2.cc

```diff
@@ -1136,15 +1140,17 @@ CheckSectionLength(const void *sectionStart, const size_t sectionLength, const v
 
 /// Checks that the area contains at least dataLength bytes after the header.
 /// The size of the field header itself is not included in dataLength.
+/// Struct FieldHeader does not contain the data itself for router_capability_header,
+/// but it does in case of router_capability_element, so specified as headerLength.
 /// \returns the total field size -- the field header and field data combined
 template<class FieldHeader>
 static size_t
-CheckFieldDataLength(const FieldHeader *header, const size_t dataLength, const void *areaStart, const size_t areaSize, const char *error)
+CheckFieldDataLength(const FieldHeader *header, const size_t headerLength, const size_t dataLength, const void *areaStart, const size_t areaSize, const char *error)
 {
     assert(header);
-    const auto dataStart = reinterpret_cast<const char*>(header) + sizeof(header);
+    const auto dataStart = reinterpret_cast<const char*>(header) + headerLength;
     CheckSectionLength(dataStart, dataLength, areaStart, areaSize, error);
-    return sizeof(header) + dataLength; // no overflow after CheckSectionLength()
+    return headerLength + dataLength; // no overflow after CheckSectionLength()
 }
```

Comment:
This code may work, but this change needs refactoring because it is a significant step backwards in the level of safety offered by this safety-focused method. It is better to break WCCP completely than to switch back to "working" WCCP with less safe code IMO. The current PR code may be 100% correct, but there is no good reason to to significantly increase the probability of future safety mistakes, similar to those we recently fixed.

This buggy function was designed to check (header, data) protocol fields or elements. In terms of the protocol, both "router capability" and "router capability element" are such elements. The official code _correctly_ (and naturally) assumes that FieldHeader is a field _header_. Converting that function parameter into a "this is a field header in some cases and something else in other cases so please try to remember to supply the correct field header length in all cases" is not a good solution.

AFAICT, the bug this PR needs to fix is in router_capability_element declaration. Here are the existing declarations that, once you strip decorations, show the problem:

```C++
struct wccp2_item_header_t { // A TLV header field
    uint16_t type;
    uint16_t length;
};

struct wccp2_capability_info_header_t { // A TLV field header
    uint16_t capability_info_type;
    uint16_t capability_info_length;
    /* dynamic length capabilities list */ // No, the list is not inside the header, but we all know what the author meant.
};

struct wccp2_capability_element_t { // A field itself!
    uint16_t capability_type;
    uint16_t capability_length;
    uint32_t capability_value;
};
```

I think we can fix the bug we are trying to fix without introducing a lot of cascading changes by adjusting the types along these lines:

```C++
struct wccp2_capability_element_header_t {
    uint16_t capability_type;
    uint16_t capability_length;
};

struct wccp2_capability_element_t: public wccp2_capability_element_header_t {
    /// convenience wrapper to access the whole header
    wccp2_capability_element_header_t *header() { return this; }
    uint32_t capability_value;
};

...

const auto elementSize = CheckFieldDataLength(
    router_capability_element.header(), ntohs(router_capability_element->capability_length),
    ...);
```

There are much better fixes, of course, but they will require a lot more code changes.

After the above is done, we should also add a static assert to the original CheckFieldDataLength(), to catch similar bugs in the future:

```C++
// A weak check that FieldType is a header rather than the whole field (or something else).
// TODO: Extract the notion of a field header into a dedicated type!
static_assert(sizeof(*header) == 4, "FieldHeader is a typical WCCPv2 TLV header");
```

After the above is done, we should also fix the sizeof(header) expressions inside the original CheckFieldDataLength(). They should read sizeof(*header) instead AFAICT. Another well-hidden bug in this untested official code!..

After the above is done, there might be at least one more bug to fix here, but I would prefer to postpone that analysis until the above changes are enacted.

I can enact these changes myself if you prefer. Just let me know.

Reply:
> I can enact these changes myself if you prefer. Just let me know.

Please do. That would be faster than explaining the details.

Reply:
Done in 542b81a.